### PR TITLE
Issue #7534: Add pmd build item

### PR DESCRIPTION
**Pull request by nmancus1**
Resolves Issue  #7810 and #7534

### CODE REVIEW

File: .ci/wercker.sh [hunk 1]
```diff
@@ -430,6 +430,17 @@ no-warning-imports-java-design-patterns)
   fi
   ;;
 
+no-error-pmd)
```
Comment (romani) [inline .ci/wercker.sh]: you need to add it to wercker.yml

to fix wercker issue, please rebase to latest master.

Comment (nmancus1) [reply .ci/wercker.sh]: Fixed

**Comment by nmancus1** [comment]
@rnveach Ok, makes a lot of sense. So, keep the `projects-to-test-on.properties` here, and open a new PR for the other two, right?  In the name of the other two, should I use the same Issue?

**Comment by rnveach** [comment]
@nmancus1 I am not sure what you are referring to. PRs need to be addressed in external repos. 
I am referring to https://github.com/checkstyle/apex-core/pull/3 and https://github.com/checkstyle/contribution/pull/457#issuecomment-596777010

**Comment by romani** [comment]
@nmancus1 , please proceed, looks like we resolved instabilities.
Please rebase and squash all commits at one.

File: wercker.yml [hunk 2]
```diff
@@ -213,17 +213,15 @@ build:
...
-  #     name: NoExceptiontest - Checkstyle ,sevntu-checkstyle javadoc
-  #     code: |
-  #       if [[ $RUN_JOB == 1 ]]; then
-  #         echo "Command: ./.ci/wercker.sh no-exception-checkstyle-sevntu-javadoc"
-  #         ./.ci/wercker.sh no-exception-checkstyle-sevntu-javadoc
-  #       else
-  #         echo "build is skipped ..."
-  #       fi
+   - script:
+       name: NoExceptiontest - Checkstyle ,sevntu-checkstyle javadoc
```
Comment (romani) [inline wercker.yml]: please move this update to some other PR or to separate commit.

Comment (nmancus1) [reply wercker.yml]: This is commit, https://github.com/checkstyle/checkstyle/pull/7816/commits/efd44b6a926c648295154bcab7875b50fc2420f3, is separate from https://github.com/checkstyle/checkstyle/pull/7816/commits/0133923af75ab4edbd49e08f6bf106c7066d8993, but would you like me to switch it to another PR?

Comment (romani) [reply wercker.yml]: This commit should not reference pmd issue 7810. Please use prefix " Issue 7534: ....", And keep it this PR

File: .ci/wercker.sh [hunk 3]
```diff
@@ -427,6 +427,17 @@ no-warning-imports-java-design-patterns)
...
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  checkout_from https://github.com/pmd/pmd.git
+  cd .ci-temp/pmd
+  mvn checkstyle:check
```
Comment (romani) [inline .ci/wercker.sh]: as the last verification .....
please modify pmd code to cause violation and confirm that it is failing.

Comment (nmancus1) [reply .ci/wercker.sh]: Is `mvn -e compile checkstyle:check -Dcheckstyle.version=${CS_POM_VERSION}` good?  It produces this:
`[ERROR] Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce-versions) on project pmd: Some Enforcer rules have failed.`

Comment (romani) [reply .ci/wercker.sh]: we expect violations from maven checkstyle plugin.

**Comment by romani** [comment]
@nmancus1 , if you would like your full name to be present in our releasenotes https://travis-ci.org/github/checkstyle/checkstyle/jobs/661935807#L583 , please update your git config to name full name, it will be used in your next commits so become populated to release notes. It is completely ok to stay with nickname.

**Comment by rnveach** [comment]
Wercker is failing horribly. It looks like there might be an issue in the YML that is breaking parsing.

**Comment by nmancus1** [comment]
> Wercker is failing horribly. It looks like there might be an issue in the YML that is breaking parsing.

I think I found it, indentation error on line 216

File: .ci/wercker.sh (hunk 3 above)
Comment (nmancus1) [reply .ci/wercker.sh]: Should build ok: https://github.com/pmd/pmd/blob/b487ef94d09855dfb3bb34f79ba267d301980b6e/pom.xml#L96

**Comment by romani** [comment]
from wercker:
```
[INFO] ----------------------< net.sourceforge.pmd:pmd >-----------------------
[INFO] Building PMD 6.23.0-SNAPSHOT                                      [1/33]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- maven-enforcer-plugin:3.0.0-M2:enforce
 (enforce-versions) @ pmd ---
[WARNING] Rule 0: org.apache.maven.plugins.enforcer.RequireJavaVersion 
failed with message:
Detected JDK Version: 1.8.0-232 is not in the allowed range [10,).
```

**Comment by romani** [comment]
wercker is failed with:
```
[INFO] --- maven-checkstyle-plugin:3.1.0:check (check) @ patch-diff-report-tool ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  8.648 s
[INFO] Finished at: 2020-03-16T22:01:38Z
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:
maven-checkstyle-plugin:3.1.0:check (check) 
on project patch-diff-report-tool: Execution check of goal 
org.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check failed: 
An API incompatibility was encountered while executing 
.apache.maven.plugins:maven-checkstyle-plugin:3.1.0:check:
 java.lang.NoSuchMethodError: com.puppycrawl.tools.checkstyle.Checker.setClassLoader(Ljava/lang/ClassLoader;)V
```

I updated https://github.com/checkstyle/contribution/blob/master/patch-diff-report-tool/pom.xml#L14 , relaunching ...

File: .ci/wercker.sh (hunk 3 above)
Comment (romani) [reply .ci/wercker.sh]: @nmancus1 , please share log (at gist.github.com) of checkstyle violation that result in build failure in pmd (please make any violation yourself).

Comment (nmancus1) [reply .ci/wercker.sh]: Just to be clear, you want me to clone pmd, make changes that will violate Checkstyle, run the script on my local pmd repo, and show a gist of Checkstyle violations?

Comment (romani) [reply .ci/wercker.sh]: yes, show whole log of execution , I need to see that it fails.

**Comment by romani** [comment]
@nmancus1 , `rmdir` can be avoided, if we use `rm -r` (without `-f`).

File: .ci/wercker.sh (hunk 3 above)
Comment (nmancus1) [reply .ci/wercker.sh]: @romani https://gist.github.com/nmancus1/eb931a5bd761d04d97cb0c66a95a3add.js

File: .ci/travis/travis.sh [hunk 4]
```diff
@@ -443,6 +443,21 @@ no-exception-test-alot-of-project1)
...
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  mkdir -p .ci-temp/
+  cd .ci-temp/
+  git clone https://github.com/pmd/pmd.git
+  cd pmd
+  mvn -e install checkstyle:check -Dcheckstyle.version=${CS_POM_VERSION}
+  find -delete
+  cd ../
+  rmdir pmd
```
Comment (romani) [inline .ci/travis/travis.sh]: please make a function for removal of folder with proptected files.

Comment (nmancus1) [reply .ci/travis/travis.sh]: @romani how would this be?
```removeFolderWithProtectedFiles() {
  while [[ "$PWD" != '/' && ${PWD##*/} != "$1" ]];
  do
    cd ..
    echo "${PWD##*/}"
  done
  if [[ "${PWD##*/}" == "$1" ]]; then
    find . -delete
    cd ..
    rmdir "$1"
    echo "$1 folder removed..."
    return 0;
  fi
  echo "unable to remove $1 folder"
}
```

Comment (romani) [reply .ci/travis/travis.sh]: I thought to make it simple
```
cd $1
find . -delete
cd ..
rmdir "$1"
```
Complicated logic in shell is very fragile and hard to support.

Comment (nmancus1) [reply .ci/travis/travis.sh]: I agree; however, some of the no-error-test projects are a few subdirectories deep like https://github.com/checkstyle/checkstyle/blob/341ef2d7296242b03bff99f37454784ccea82906/.ci/wercker.sh#L156 and https://github.com/checkstyle/checkstyle/blob/341ef2d7296242b03bff99f37454784ccea82906/.ci/wercker.sh#L169.  I felt the recursive check for correct directory would be necessary for those like this.

Comment (romani) [reply .ci/travis/travis.sh]: .git folder is always at same place, it is completely different fact that we in repo use different folder for validation. 
We created repo folder we should clean up it. We should not care why we did inside repo.

Comment (romani) [reply .ci/travis/travis.sh]: Let's try simple first and improve later on

Comment (nmancus1) [reply .ci/travis/travis.sh]: Fixed.

File: .ci/travis/travis.sh [hunk 5]
```diff
@@ -443,6 +449,19 @@ no-exception-test-alot-of-project1)
...
+no-error-pmd)
+  CS_POM_VERSION=$(mvn -e -q -Dexec.executable='echo' -Dexec.args='${project.version}' \
+                     --non-recursive org.codehaus.mojo:exec-maven-plugin:1.3.1:exec)
+  echo CS_version: ${CS_POM_VERSION}
+  mkdir -p .ci-temp/
+  cd .ci-temp/
+  git clone https://github.com/pmd/pmd.git
+  cd pmd
+  mvn -e install checkstyle:check -Dcheckstyle.version=${CS_POM_VERSION}
+  removeFolderWithProtectedFiles pmd
```
Comment (romani) [inline .ci/travis/travis.sh]: Please do 
`cd ..`
Before this line, to remove folder from top, not from inside

File: .ci/wercker.sh (hunk 3 above)
Comment (romani) [reply .ci/wercker.sh]: ok, looks good.
```
[INFO] --- maven-checkstyle-plugin:3.1.1:check (checkstyle-check)
 @ pmd-core ---
[INFO] There is 1 error reported by Checkstyle 8.31-SNAPSHOT 
with /net/sourceforge/pmd/pmd-checkstyle-config.xml ruleset.
[ERROR] src/main/java/net/sourceforge/pmd/lang/ast/Node.java:[1]
 (misc) NewlineAtEndOfFile: File does not end with a newline.
[INFO] ------------------------------------------------------------------------

```

File: .ci/travis/travis.sh (hunk 5 above)
Comment (romani) [reply .ci/travis/travis.sh]: this is still not done.

File: .ci/travis/travis.sh [hunk 6]
```diff
@@ -2,6 +2,12 @@
 # Attention, there is no "-x" to avoid problems on Travis
 set -e
 
+removeFolderWithProtectedFiles() {
+  find . -delete
```
Comment (romani) [inline .ci/travis/travis.sh]: please do `cd "$1"` to do deletion inside target folder.

Comment (nmancus1) [reply .ci/travis/travis.sh]: I'm curious; why `cd ..` then `cd "$1"`?  To make sure we're in the right directory?

Comment (romani) [reply .ci/travis/travis.sh]: yes, and you raise a good point we should do `cd -`  at the end to return back to where we were before function execution.

Comment (romani) [reply .ci/travis/travis.sh]: Please send separate PR to add `cd -`.
I am merging this PR sooner as another PR depends on it.

**Pull request merged**